numba/cuda/cudadrv/driver.py

```diff
@@ -894,12 +894,28 @@ def unload_module(self, module):
     def get_default_stream(self):
         return Stream(weakref.proxy(self), drvapi.cu_stream(0), None)
 
+    def get_legacy_default_stream(self):
+        handle = drvapi.cu_stream(drvapi.CU_STREAM_LEGACY)
+        return Stream(weakref.proxy(self), handle, None)
+
+    def get_per_thread_default_stream(self):
+        handle = drvapi.cu_stream(drvapi.CU_STREAM_PER_THREAD)
+        return Stream(weakref.proxy(self), handle, None)
+
     def create_stream(self):
         handle = drvapi.cu_stream()
         driver.cuStreamCreate(byref(handle), 0)
         return Stream(weakref.proxy(self), handle,
                       _stream_finalizer(self.deallocations, handle))
 
+    def create_external_stream(self, ptr):
+        if not isinstance(ptr, int):
+            raise TypeError("ptr for external stream must be an int")
+        handle = drvapi.cu_stream(ptr)
+        return Stream(weakref.proxy(self), handle,
+                      _stream_finalizer(self.deallocations, handle),
```

Comment:
Just realised this line is wrong - external streams should not have a Numba stream finalizer, because it will try to destroy them even though it doesn't own them. Will fix...

Reply:
I thought this line means when the CUDA context is about to be destroyed, the stream will be destroyed? I suppose this should be OK since the external library will share the same CUDA context with Numba?

Reply:
The external library (e.g. CuPy) will probably want to call `cuStreamDestroy` on it - with the `_stream_finalizer`, Numba will also call `cuStreamDestroy`, which results in the `CUDA_ERROR_INVALID_HANDLE` error.

Reply:
In other words - the user will be managing the lifetime of the stream, and Numba shouldn't be.

Reply:
Ah I see...Don't know what I was thinking 😅 Good catch, Graham!

Reply:
Thanks... Though I was originally made aware of the issue because I pointed someone to https://gist.github.com/leofang/4a043e5d94b4702d04fde2b9e7dcebbd who used `stream_cupy_to_numba` in their code and found that the double-deletion of streams occurring. They removed the finalizer ( https://gist.github.com/leofang/4a043e5d94b4702d04fde2b9e7dcebbd#file-stream_cupy_to_numba-py-L28 ) and then the error was resolved.

Reply:
Ahhh sorry for the buggy gist 🤦 I am scratching my head and wondering why I didn't encounter this bug so far...

Reply:
One possibility is that the Numba finalizer could have only been called at exit time in your scenario, which would result in a traceback being printed, but wouldn't cause abnormal program termination - could that be what's happening?